eth/tracers/api.go

```diff
@@ -682,86 +768,83 @@ func (api *API) traceInternalsAndAccounts(ctx context.Context, block *types.Bloc
+		results   = &internalAndAccountResult{
 			InternalTxs:   make([]*txTraceResult, len(txs)),
 			DirtyAccounts: make([]*dirtyAccount, 0),
 		}
-
-		pend = new(sync.WaitGroup)
-		jobs = make(chan *txTraceTask, len(txs))
+		pend   sync.WaitGroup
+		failed error
 	)
-	threads := runtime.NumCPU()
-	if threads > len(txs) {
-		threads = len(txs)
-	}
-	blockCtx := core.NewEVMBlockContext(block.Header(), api.chainContext(ctx), nil)
-	blockHash := block.Hash()
+
 	for th := 0; th < threads; th++ {
 		pend.Add(1)
 		go func() {
 			defer pend.Done()
 			// Fetch and execute the next transaction trace tasks
 			for task := range jobs {
 				msg, _ := txs[task.index].AsMessage(signer, block.BaseFee())
-				txHash := txs[task.index].Hash()
 				txctx := &Context{
-					BlockHash: blockHash,
-					TxIndex:   task.index,
-					TxHash:    txHash,
+					BlockHash:   blockHash,
+					BlockNumber: block.Number(),
+					TxIndex:     task.index,
+					TxHash:      txs[task.index].Hash(),
 				}
 				res, err := api.traceTx(ctx, msg, txctx, blockCtx, task.statedb, config, block)
 				if err != nil {
-					results.InternalTxs[task.index] = &txTraceResult{
-						TransactionHash: txHash,
-						Error:           err.Error(),
-					}
+					results.InternalTxs[task.index] = &txTraceResult{TransactionHash: txs[task.index].Hash(), Error: err.Error()}
 					continue
 				}
-				results.InternalTxs[task.index] = &txTraceResult{
-					TransactionHash: txHash,
-					Result:          res,
-				}
+				results.InternalTxs[task.index] = &txTraceResult{TransactionHash: txs[task.index].Hash(), Result: res}
 			}
 		}()
 	}
-	// Feed the transactions into the tracers and return
-	var failed error
+txloop:
 	for i, tx := range txs {
 		// Send the trace task over for execution
-		jobs <- &txTraceTask{statedb: statedb.Copy(), index: i}
+		task := &txTraceTask{statedb: statedb.Copy(), index: i}
+		select {
+		case <-ctx.Done():
+			failed = ctx.Err()
+			break txloop
+		case jobs <- task:
+		}
 
 		// Generate the next state snapshot fast without tracing
 		msg, _ := tx.AsMessage(signer, block.BaseFee())
 		statedb.Prepare(tx.Hash(), i)
-		vmenv := vm.NewEVM(blockCtx, core.NewEVMTxContext(msg), statedb, api.backend.ChainConfig(), vm.Config{})
 		consortium.HandleSubmitBlockReward(api.backend.Engine(), statedb, msg, block)
+		vmenv := vm.NewEVM(blockCtx, core.NewEVMTxContext(msg), statedb, api.backend.ChainConfig(), vm.Config{})
 		if _, err := core.ApplyMessage(vmenv, msg, new(core.GasPool).AddGas(msg.Gas())); err != nil {
 			failed = err
-			break
+			break txloop
 		}
 		// Finalize the state so any modifications are written to the trie
 		// Only delete empty objects if EIP158/161 (a.k.a Spurious Dragon) is in effect
-		statedb.Finalise(vmenv.ChainConfig().IsEIP158(block.Number()))
+		statedb.Finalise(is158)
 	}
+
 	close(jobs)
 	pend.Wait()
 
-	// Computes the current root hash of the state trie
-	statedb.IntermediateRoot(api.backend.ChainConfig().IsEIP158(block.Number()))
-
```

Comment:
We must not remove these lines. Dirty account has a field to store the root of the trie, this `IntermediateRoot` call helps to calculate the root of the trie

Reply:
I see